### PR TITLE
fix: minimize number of adapter objects in cache

### PR DESCRIPTION
**Pull request by siillee**
## Description

This PR fixes an issue we had where multiple adapter objects were being created for the same vector and graph databases. 
We should have one adapter object per database (i.e. per dataset), and the fixes in this PR ensure that this is true.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Code refactoring
- [ ] Other (please specify):

## Pre-submission Checklist

- [ ] **I have tested my changes thoroughly before submitting this PR** (See `CONTRIBUTING.md`)
- [ ] **This PR contains minimal changes necessary to address the issue/feature**
- [ ] My code follows the project's coding standards and style guidelines
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if applicable)
- [ ] All new and existing tests pass
- [ ] I have searched existing PRs to ensure this change hasn't been submitted already
- [ ] I have linked any relevant issues in the description
- [ ] My commits have clear and descriptive messages

## DCO Affirmation
I affirm that all code in every commit of this pull request conforms to the terms of the Topoteretes Developer Certificate of Origin.

## Summary by CodeRabbit

* **Refactor**
  * Improved handling and normalization of optional database engine parameters, including unified defaulting and coercion of port values and environment-driven defaults.
  * Ensured normalized values are used consistently when selecting provider paths, improving reliability of engine initialization and adapter caching.

* **Tests**
  * Added an async validation test that exercises dataset ingestion, search, and verifies adapter cache sizes remain within expected limits.

### CODE REVIEW

**Review by hajdul88 — COMMENTED** [review]
There are still some edge cases, especially in runtime-configured setups, where multiple adapter objects can still be created. That said, it improves the behavior for the main configuration paths and should solve the issue in most practical cases.

I’d still like to see at least a simple test added so we preserve this normalization behavior going forward and don’t accidentally change it later.

File: cognee/infrastructure/databases/vector/create_vector_engine.py [hunk 1]
```diff
@@ -1,13 +1,52 @@
...
 
 from .supported_databases import supported_databases
 from .embeddings import get_embedding_engine
-from cognee.infrastructure.databases.graph.config import get_graph_context_config
 
 from functools import lru_cache
 from cognee.shared.lru_cache import DATABASE_MAX_LRU_CACHE_SIZE
 
 
+@lru_cache(maxsize=1)
```
Comment (dexters1) [inline cognee/infrastructure/databases/vector/create_vector_engine.py]: Remove cache

File: cognee/infrastructure/databases/graph/get_graph_engine.py [hunk 2]
```diff
@@ -8,6 +12,43 @@
 from .supported_databases import supported_databases
 
 
+@lru_cache(maxsize=1)
```
Comment (dexters1) [inline cognee/infrastructure/databases/graph/get_graph_engine.py]: Delete this decorator too

File: cognee/infrastructure/databases/vector/create_vector_engine.py [hunk 3]
```diff
@@ -24,7 +63,14 @@ def create_vector_engine(
...
+
+    normalized_optional_params = _normalize_optional_create_vector_engine_params(locals())
+    vector_db_port = normalized_optional_params["vector_db_port"]
+    vector_db_key = normalized_optional_params["vector_db_key"]
+    vector_dataset_database_handler = normalized_optional_params["vector_dataset_database_handler"]
+    vector_db_username = normalized_optional_params["vector_db_username"]
+    vector_db_password = normalized_optional_params["vector_db_password"]
+    vector_db_host = normalized_optional_params["vector_db_host"]
 
     unified_provider = os.environ.get("USE_UNIFIED_PROVIDER", "")
```
Comment (dexters1) [inline cognee/infrastructure/databases/vector/create_vector_engine.py:74]: Can we move this to the inner _create so it works properly? In both graph and vector engines

Comment (siillee) [reply cognee/infrastructure/databases/vector/create_vector_engine.py:74]: Seems that the env check was moved outside of cache on purpose:
https://github.com/topoteretes/cognee/commit/606e57fa5cdd9d3bc70f4630618b38728e3acb66

I left the env check, and passed the `unified_provider`, so that I can move the creation of the adatper to the private function.

EDIT: Separate PR will handle these changes

**Comment by hajdul88** [comment]
Agreed to merge this, issues raised by Igor were not directly connected to this PR so they were addressed in another PR

**Review by hajdul88 — APPROVED** [review]
